### PR TITLE
Added post for Basic Authentication custom authorizer

### PR DESCRIPTION
**Pull request by svdgraaf**
127 0 0 1_4000_2018_06_13_serverless-basic-authentication-custom-authorizer html 1

### CODE REVIEW

File: _posts/2018-06-13-Serverless-Basic-Authentication-Custom-Authorizer.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+layout: post
+title: Serverless Basic Authentication using a Custom Authorizer
+banner: /assets/posts/2018-06-13-Serverless-Basic-Authentication-Custom-Authorizer/banner.png
+github: svdgraaf/serverless-basic-authentication
+author: svdgraaf
+---
+
+In a recent project, we needed our api's to be able to work with external systems. These systems only supported http basic authentication (eg: username/password) for integrating with external systems.
+
+Of course, Basic HTTP Authentication is the easiest and most straight forward way to integrate things and comes build in, in almost any http client.
```
Comment (donkersgoed) [inline _posts/2018-06-13-Serverless-Basic-Authentication-Custom-Authorizer.md]: "built in",

Suggestion for readability: 
"Of course, Basic HTTP Authentication is the easiest and most straight forward way to integrate things, and it's built into almost any http client.

File: _posts/2018-06-13-Serverless-Basic-Authentication-Custom-Authorizer.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+If you want to read more about basic authentication, I suggest you take a look at the [wikipedia page](https://en.wikipedia.org/wiki/Basic_access_authentication) or [RFC7617](https://tools.ietf.org/html/rfc7617).
+
+Api Gateway
+-----------
+HTTP Basic Authentication is a very simple standard, which is why almost everyone supports it. Except for Api Gateway. The easiest way to setup authentication on Api Gateway is either to use Api Keys or IAM authentication. Using IAM authentication makes you have to sign all requests. Using the Api Keys, you need to add a header (`x-api-key` to each request, with the api key in it).
```
Comment (donkersgoed) [inline _posts/2018-06-13-Serverless-Basic-Authentication-Custom-Authorizer.md]: "HTTP Basic Authentication is a very simple standard, which is why almost everyone supports it. Except for Api Gateway."

->

"HTTP Basic Authentication is a very simple standard, which is why almost everyone supports it. But not Api Gateway."

File: _posts/2018-06-13-Serverless-Basic-Authentication-Custom-Authorizer.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
+---
+layout: post
+title: Serverless Basic Authentication using a Custom Authorizer
+banner: /assets/posts/2018-06-13-Serverless-Basic-Authentication-Custom-Authorizer/banner.png
+github: svdgraaf/serverless-basic-authentication
+author: svdgraaf
+---
+
+In a recent project, we needed our api's to be able to work with external systems. These systems only supported http basic authentication (eg: username/password) for integrating with external systems.
```
Comment (donkersgoed) [inline _posts/2018-06-13-Serverless-Basic-Authentication-Custom-Authorizer.md]: General remark: always write "API" and "HTTP" in full upper case.

File: _posts/2018-06-13-Serverless-Basic-Authentication-Custom-Authorizer.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+-----------
+HTTP Basic Authentication is a very simple standard, which is why almost everyone supports it. Except for Api Gateway. The easiest way to setup authentication on Api Gateway is either to use Api Keys or IAM authentication. Using IAM authentication makes you have to sign all requests. Using the Api Keys, you need to add a header (`x-api-key` to each request, with the api key in it).
+
+Fortunately, nowadays Api Gateway supports custom authorizers. Also, it is now possible to configure Api Gateway to get the api key from the custom authorizer. Which is everything we need to support basic authentication.
+
+We have the user provide us with the Basic Authentication header (called `Authorization`). As a username, they pass the api key name, and as the password, they send the api key value. The client (browser) will do some magiuc (base64 encoding the username/password) and sends it along. We then pass the value of this header to the custom authorizer, which base64 decodes it, checks if the api key is valid, and forwards the api key value back to Api Gateway. Api Gateway will then use that api key and forward the context to the receiving endpoint (eg: Lambda function).
```
Comment (donkersgoed) [inline _posts/2018-06-13-Serverless-Basic-Authentication-Custom-Authorizer.md]: magiuc

File: _posts/2018-06-13-Serverless-Basic-Authentication-Custom-Authorizer.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                'Resource': arn
+            }]
+        }
+    }
+    print("Authentication response: %s" % authResponse)
+
+    return authResponse
+```
+
+The real magic is in the response code. In the AWS documentation this is fairly poorly documented but the magic is in the `usageIdentifierKey` return value. When this key is present. And Api Gateway is configured to get the key value from the custom authorizer, Api Gateway will use the value of this key as the api key. It needs to contain the *value* of the key, not the name, id or whatever, it's the actual value of the key (took me a while to figure that one out).
```
Comment (donkersgoed) [inline _posts/2018-06-13-Serverless-Basic-Authentication-Custom-Authorizer.md]: "When this key is present. And Api Gateway"

-> 

"When this key is present and Api Gateway"

**Comment by svdgraaf** [comment]
@donkersgoed thanks, all fixed :)

**Pull request merged**